cms/search/publishers.py

```diff
@@ -0,0 +1,176 @@
+    for created/updated or deleted events. Each subclass only needs to define how
+    to actually publish (i.e., send) the built message.
+
+    dp-search-data-extractor spec link:
+    https://github.com/ONSdigital/dp-search-data-extractor/blob/develop/specification.yml#L53
+    dp-search-data-importer spec link:
+    https://github.com/ONSdigital/dp-search-data-importer/blob/30fb507e90f2cf1974ec0ca43bb0466307e2f112/specification.yml#L186
+    contract: https://github.com/ONSdigital/dis-search-upstream-stub/blob/main/docs/contract/resource_metadata.yml
+    """
+
+    def publish_created_or_updated(self, page: "Page") -> None:
+        """Build the message for the created/updated event.
+        Delegate sending to the subclass's _publish().
+        """
+        channel = self.get_created_or_updated_channel()
+        message = self._construct_message_for_create_update(page)
+        logger.info(
+            "BasePublisher: About to publish created/updated message=%s to channel=%s",
+            message,
+            channel,
+        )
+        return self._publish(channel, message)
+
+    def publish_deleted(self, page: "Page") -> None:
+        """Build the message for the deleted event.
+        Delegate sending to the subclass's _publish().
+        """
+        channel = self.get_deleted_channel()
+        message = {
+            "uri": page.url_path,
+        }
+        logger.info(
+            "BasePublisher: About to publish deleted message=%s to channel=%s",
+            message,
+            channel,
+        )
+        return self._publish(channel, message)
+
+    @abstractmethod
+    def _publish(self, channel: str, message: dict) -> None:
+        """Each child class defines how to actually send/publish
+        the message (e.g., Kafka, logging, etc.).
+        """
+
+    @abstractmethod
+    def get_created_or_updated_channel(self) -> str:
+        """Provide the channel (or other necessary routing key) for created/updated.
+        This can be a no-op or empty string for some implementations.
+        """
+
+    @abstractmethod
+    def get_deleted_channel(self) -> str:
+        """Provide the channel (or other necessary routing key) for deleted messages.
+        This can be a no-op or empty string for some implementations.
+        """
+
+    def _construct_message_for_create_update(self, page: "Page") -> dict:
+        """Build a dict that matches the agreed metadata schema for 'created/updated'.
+
+        resource_metadata.yml:
+           - StandardPayload: requires {uri, title, content_type}, optional fields
+           - ReleasePayload: extends StandardPayload with release-specific fields
+        """
+        # Common fields for StandardPayload (also part of ReleasePayload)
+        # The schema requires at minimum: uri, title, content_type
+        message = {
+            "uri": page.url_path,
+            "content_type": page.search_index_content_type,
```

Comment:
suggestion: either use `getattr()` with a generic defualt, or add an attribute on `BasePage` with a generic default. Or add a Django system check that if someone adds a new page type to the list to send to the search index, it has a `search_index_content_type` attribute.

Reply:
Went with the Django system check!